Using kcadm.sh create components result to 400 Bad Request

### Before reporting an issue

- [x] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

admin/cli

### Describe the bug

I'm using Keycloak 23.0.3 in Docker. After launch the container and the container is ready, I authenticate as admin, using `kcadm.sh config credentials --server http://localhost:$KC_INTERNAL_PORT --user $KEYCLOAK_ADMIN --password $KEYCLOAK_ADMIN_PASSWORD --realm master`. After that, I create a realm being my-realm, and I want to create a component using `kcadm.sh create components -r my-realm -s name=my-realm-component -s parentId=my-realm -s providerId=my-component-id -s providerType=my.package.providerType`. This request gives me a 400 Bad Request.
With the parameter -x, the stacktrace is :
`null [HTTP 400 Bad Request]
org.keycloak.client.admin.cli.util.HttpResponseException: null [HTTP 400 Bad Request]
        at org.keycloak.client.admin.cli.util.HeadersBodyStatus.checkSuccess(HeadersBodyStatus.java:61)
        at org.keycloak.client.admin.cli.util.HttpUtil.checkSuccess(HttpUtil.java:353)
        at org.keycloak.client.admin.cli.commands.AbstractRequestCmd.process(AbstractRequestCmd.java:365)
        at org.keycloak.client.admin.cli.commands.AbstractRequestCmd.execute(AbstractRequestCmd.java:128)
        at org.jboss.aesh.console.command.container.DefaultCommandContainer.executeCommand(DefaultCommandContainer.java:63)
        at org.jboss.aesh.console.command.container.DefaultCommandContainer.executeCommand(DefaultCommandContainer.java:48)
        at org.keycloak.client.admin.cli.aesh.AeshConsoleCallbackImpl.execute(AeshConsoleCallbackImpl.java:54)
        at org.jboss.aesh.console.AeshProcess.run(AeshProcess.java:53)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.lang.RuntimeException: {"error":"HTTP 400 Bad Request"}
        ... 11 more`

### Version

23.0.3

### Expected behavior

Creates the components and returning a 200 Success.

### Actual behavior

Do not create the component and returns a 400 Bad Request.

### How to Reproduce?

Building image quay.io/keycloak/keycloak:23.0.3 with
`ENV KEYCLOAK_ADMIN=admin
ENV KEYCLOAK_ADMIN_PASSWORD=admin
ENTRYPOINT["/opt/keycloak/bin/kc.sh"]
CMD["start-dev"]`

Launches the conainer.
Waiting the container is ready.
Using the container terminal.
Executing `cd /opt/keycloak/bin`
Executing `./kcadm.sh config credentials --server http://localhost:8080 --user admin --password admin --realm master`
Executing `./kcadm.sh create realms -s realm=my-realm -s enabled=true`
Executing `./kcadm.sh create components -r my-realm -s name=my-realm-component -s parentId=my-realm -s providerId=my-component-id -s providerType=my.package.providerType`
Gets a 400 Bad Request.

### Anything else?

_No response_

If you run the server with --log-level=debug you may see:

```
2023-12-21 15:33:34,491 DEBUG [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-1) Error response 400: jakarta.ws.rs.BadRequestException: HTTP 400 Bad Request
        at org.keycloak.services.resources.admin.ComponentResource.create(ComponentResource.java:148)
        at org.keycloak.services.resources.admin.ComponentResource$quarkusrestinvoker$create_6ceaf3ab48a6cd738c575b56e05fde2f99766f34.invoke(Unknown Source)
        at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
        at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:145)
        at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:576)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.IllegalArgumentException: Invalid provider type 'my.package.providerType'
        at org.keycloak.models.utils.ComponentUtil.getComponentFactory(ComponentUtil.java:82)
        at org.keycloak.models.utils.ComponentUtil.getComponentFactory(ComponentUtil.java:57)
        at org.keycloak.models.jpa.RealmAdapter.importComponentModel(RealmAdapter.java:2032)
        at org.keycloak.models.jpa.RealmAdapter.addComponentModel(RealmAdapter.java:2016)
        at org.keycloak.models.cache.infinispan.RealmAdapter.addComponentModel(RealmAdapter.java:1519)
        at org.keycloak.services.resources.admin.ComponentResource.create(ComponentResource.java:141)
        ... 11 more
```

Can you use debug to see what exactly is the problem on the server side?

Ideally the exception message should be provided to the admin client in some form, unforutnately all that is conveyed is the bad request message. This issue should at least cover improving that.

--- a/services/src/main/java/org/keycloak/services/error/KeycloakErrorHandler.java
+++ b/services/src/main/java/org/keycloak/services/error/KeycloakErrorHandler.java
@@ -85,7 +85,8 @@ public class KeycloakErrorHandler implements ExceptionMapper<Throwable> {
             OAuth2ErrorRepresentation error = new OAuth2ErrorRepresentation();
 
             error.setError(getErrorCode(throwable));
-            
+            error.setErrorDescription("For more on this error consult the server log at the debug level.");
+
             return Response.status(statusCode)
                     .header(HttpHeaders.CONTENT_TYPE, jakarta.ws.rs.core.MediaType.APPLICATION_JSON_TYPE.toString())
                     .entity(error)
@@ -125,11 +126,11 @@ public class KeycloakErrorHandler implements ExceptionMapper<Throwable> {
         if (throwable instanceof JsonProcessingException) {
             status = Response.Status.BAD_REQUEST.getStatusCode();
         }
-        
+
         if (throwable instanceof ModelDuplicateException) {
             status = Response.Status.CONFLICT.getStatusCode();
         }
-        
+
         return status;
     }
 

--- a/services/src/main/java/org/keycloak/services/resources/admin/ClientAttributeCertificateResource.java
+++ b/services/src/main/java/org/keycloak/services/resources/admin/ClientAttributeCertificateResource.java
@@ -197,7 +197,9 @@ public class ClientAttributeCertificateResource {
         CertificateRepresentation info = new CertificateRepresentation();
         MultivaluedMap<String, FormPartValue> uploadForm = session.getContext().getHttpRequest().getMultiPartFormParameters();
         FormPartValue keystoreFormatPart = uploadForm.getFirst("keystoreFormat");
-        if (keystoreFormatPart == null) throw new BadRequestException();
+        if (keystoreFormatPart == null) {
+            throw new BadRequestException("keystoreFormat cannot be null");
+        }
         String keystoreFormat = keystoreFormatPart.asString();
         FormPartValue inputParts = uploadForm.getFirst("file");
         if (keystoreFormat.equals(CERTIFICATE_PEM)) {

--- a/services/src/main/java/org/keycloak/services/resources/admin/ComponentResource.java
+++ b/services/src/main/java/org/keycloak/services/resources/admin/ComponentResource.java
@@ -145,7 +145,7 @@ public class ComponentResource {
         } catch (ComponentValidationException e) {
             return localizedErrorResponse(e);
         } catch (IllegalArgumentException e) {
-            throw new BadRequestException(e);
+            throw new BadRequestException("Invalid provider type or no such provider", e);
         }
     }
 
@@ -184,7 +184,7 @@ public class ComponentResource {
         } catch (ComponentValidationException e) {
             return localizedErrorResponse(e);
         } catch (IllegalArgumentException e) {
-            throw new BadRequestException();
+            throw new BadRequestException("Invalid provider type or no such provider", e);
         }
     }
     @DELETE

--- a/services/src/main/java/org/keycloak/services/resources/admin/RoleContainerResource.java
+++ b/services/src/main/java/org/keycloak/services/resources/admin/RoleContainerResource.java
@@ -143,7 +143,7 @@ public class RoleContainerResource extends RoleResource {
         auth.roles().requireManage(roleContainer);
 
         if (rep.getName() == null) {
-            throw new BadRequestException();
+            throw new BadRequestException("role has no name");
         }
 
         try {

--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/error/UncaughtErrorPageTest.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/error/UncaughtErrorPageTest.java
@@ -26,7 +26,6 @@ import org.keycloak.util.JsonSerialization;
 import org.keycloak.utils.MediaType;
 import org.openqa.selenium.By;
 
-import jakarta.ws.rs.core.Response;
 import java.io.IOException;
 import java.io.InputStream;
 import java.net.MalformedURLException;
@@ -38,10 +37,13 @@ import java.util.HashSet;
 import java.util.List;
 import java.util.Objects;
 
+import jakarta.ws.rs.core.Response;
+
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.notNullValue;
 import static org.hamcrest.core.Is.is;
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertTrue;
 import static org.keycloak.utils.MediaType.APPLICATION_JSON;
@@ -95,7 +97,7 @@ public class UncaughtErrorPageTest extends AbstractKeycloakTest {
 
             OAuth2ErrorRepresentation error = JsonSerialization.readValue(response.getEntity().getContent(), OAuth2ErrorRepresentation.class);
             assertEquals(OAuthErrorException.INVALID_REQUEST, error.getError());
-            assertNull(error.getErrorDescription());
+            assertNotNull(error.getErrorDescription());
         }
     }
 
@@ -115,7 +117,7 @@ public class UncaughtErrorPageTest extends AbstractKeycloakTest {
 
             OAuth2ErrorRepresentation error = JsonSerialization.readValue(response.getEntity().getContent(), OAuth2ErrorRepresentation.class);
             assertEquals(OAuthErrorException.INVALID_REQUEST, error.getError());
-            assertNull(error.getErrorDescription());
+            assertNotNull(error.getErrorDescription());
         }
     }
 

